File: core/core-domain/src/main/java/com/nmh/commerce/coupon/CouponIssuer.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.user.User;
import com.nmh.commerce.user.UserRepository;
import com.nmh.commerce.utils.LocalDateTimeHolder;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Component
public class CouponIssuer {
    private final CouponRepository couponRepository;
    private final IssuedCouponRepository issuedCouponRepository;
    private final UserRepository userRepository;
    private final CouponStockManager stockManager;
    private final LocalDateTimeHolder localDateTimeHolder;

    @Transactional
    public IssuedCoupon issue(Long couponId, Long userId) {
        Coupon coupon = couponRepository.findById(couponId).orElseThrow(() -> new IllegalArgumentException("해당 쿠폰을 찾을 수 없습니다."));
        User user = userRepository.findById(userId).orElseThrow(() -> new IllegalArgumentException("해당 유저를 찾을 수 없습니다."));
        IssuedCoupon issuedCoupon = IssuedCoupon.issue(coupon, user, localDateTimeHolder.now());
        stockManager.deductStock(couponId);
        return issuedCouponRepository.save(issuedCoupon);
    }
}

File: core/core-application/src/main/kotlin/com/nmh/commerce/coupon/CouponIssuer.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.coupon.IssuedCoupon.Companion.issue
import com.nmh.commerce.user.UserRepository
import com.nmh.commerce.utils.LocalDateTimeHolder
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.function.Supplier

@Component
class CouponIssuer(
    private val couponRepository: CouponRepository,
    private val issuedCouponRepository: IssuedCouponRepository,
    private val userRepository: UserRepository,
    private val stockManager: CouponStockManager,
    private val localDateTimeHolder: LocalDateTimeHolder
) {
    @Transactional
    fun issue(couponId: Long, userId: Long): IssuedCoupon {
        val coupon = couponRepository.findById(couponId)
        val user = userRepository.findById(userId)
        val issuedCoupon = issue(coupon, user, localDateTimeHolder.now())
        stockManager.deductStock(couponId)
        return issuedCouponRepository.save(issuedCoupon)
    }
}
